src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -164,10 +164,21 @@ internal Task Ensure(int count, bool async, bool readingNotifications)
                     buffer.ReadPosition = 0;
                 }
 
-                var finalCt = async && buffer.Timeout != TimeSpan.Zero
+                var finalCt = buffer.Timeout != TimeSpan.Zero
                     ? buffer.Cts.Start()
                     : buffer.Cts.Reset();
 
+                CancellationTokenRegistration syncCancellationRegistration = default;
+                if (!async)
+                {
+                    syncCancellationRegistration = finalCt.Register(conn =>
```

Comment:
```suggestion
                    // In sync I/O, the token isn't passed to the read, so we use a registration forcefully break the connector.
                    // This will cause the read operation to throw an exception which will be handled below.
                    syncCancellationRegistration = finalCt.Register(conn =>
```